src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistUtils.java

```diff
@@ -0,0 +1,58 @@
+package frc.robot.subsystems.swerve.swervestatehelpers;
+
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.ChassisSpeeds;
+import frc.robot.subsystems.swerve.SwerveConstants;
+
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class AimAssistUtils {
+
+    /**
+     * @param speeds the robot speeds at "field-relative"
+     *
+     * @return the magnitude of the drive velocities vecotr
+     * */
+    private static double getDriveMagnitude(ChassisSpeeds speeds) {
+        return Math.sqrt(Math.pow(speeds.vxMetersPerSecond, 2) + Math.pow(speeds.vyMetersPerSecond, 2));
+    }
+
+    /**
+     * @param currentSpeeds current chassis speeds, field relative
+     * @param inputSpeeds the speeds that the joysticks command on the robot
+     * @param robotPoseSupplier supplier of the robot position
+     * @param targetRotationSupplier target angle for the rotation (as a supplier)
+     * @param swerveConstants the constants of the swerve
+     *
+     * @return the aim assisted chassis speeds with only rotation speeds assisted (to look at somewhere)
+     *
+     * */
+
+    public static ChassisSpeeds getRotationAssistedSpeeds(ChassisSpeeds inputSpeeds, ChassisSpeeds currentSpeeds,
+            Supplier<Pose2d> robotPoseSupplier,
+            Function<Pose2d, Rotation2d> targetRotationSupplier, SwerveConstants swerveConstants) {
+
+        double driveMagnitude = getDriveMagnitude(currentSpeeds);
+        Rotation2d pidVelocity = Rotation2d.fromDegrees(swerveConstants.rotationDegreesPIDController().calculate(
+                robotPoseSupplier.get().getRotation().getDegrees(),
+                targetRotationSupplier.apply(robotPoseSupplier.get()).getDegrees()
+        ));
+
+        double angularVelocityRads =
+                pidVelocity.getRadians() * SwerveConstants.AIM_ASSIST_MAGNITUDE_FACTOR / (driveMagnitude + SwerveConstants.AIM_ASSIST_MAGNITUDE_FACTOR);
+
+        double angularVelocityWithJoystick = angularVelocityRads + inputSpeeds.omegaRadiansPerSecond;
+
+        double clampedAngularVelocity = MathUtil.clamp(
+                angularVelocityWithJoystick,
+                -swerveConstants.maxRotationalVelocityPerSecond().getRadians(),
+                swerveConstants.maxRotationalVelocityPerSecond().getRadians()
+        );
+
+        return new ChassisSpeeds(inputSpeeds.vxMetersPerSecond, inputSpeeds.vyMetersPerSecond, clampedAngularVelocity);
+    }
```

Comment:
this is a huge function, move math stuff to swerve math, and split to functions

Reply:
its long because of line seperation. this whole function can be 5 lines if we will shorten it. most of the stuff is not math stuff but PID calculations, normalization stuff and so on. I'll try moving but I can but it wont shorten it for much

Reply:
So split to functions and keep it longggggggggggggggg not like your something😉

Reply:
i did my best, if there is anything more comment later